File: go/classifier.go
package leanprompt

// Content-type classifier + RepeatTracker (parity-spec §6).

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

// ContentType labels emitted by the classifier; string values are
// config-visible and frozen.
type ContentType string

const (
	Unknown       ContentType = "unknown"
	Prose         ContentType = "prose"
	Code          ContentType = "code"
	ErrorType     ContentType = "error"
	Structured    ContentType = "structured"
	Repeat        ContentType = "repeat"
	LongImportant ContentType = "long_important"
)

func contentTypeFrom(s string) (ContentType, bool) {
	switch ContentType(s) {
	case Unknown, Prose, Code, ErrorType, Structured, Repeat, LongImportant:
		return ContentType(s), true
	}
	return "", false
}

var errorMarkers = []string{
	"Traceback (most recent call last):",
	"Uncaught exception",
	"UnhandledPromiseRejection",
	"thread 'main' panicked at",
	"panic: ",
	"Exception in thread",
	"FATAL: ",
	"ERROR: ",
	"Error: ",
	"Exception: ",
	"java.lang.",
}

var codeLinePrefixes = []string{
	"def ", "class ", "function ", "async function ", "import ", "from ",
	"export ", "package ", "#include", "fn ", "pub fn ", "func ", "var ",
	"const ", "let ",
}

const (
	minCodeLines       = 2
	structuredMinChars = 200
)

// Classify a message: ERROR > CODE > STRUCTURED > PROSE; UNKNOWN when the
// extracted text is empty.
func Classify(message Message) ContentType {
	text := GetTextContent(message)
	if strings.TrimSpace(text) == "" {
		return Unknown
	}
	if looksLikeError(text) {
		return ErrorType
	}
	if looksLikeCode(text) {
		return Code
	}
	if looksLikeStructured(text) {
		return Structured
	}
	return Prose
}

func looksLikeError(text string) bool {
	for _, m := range errorMarkers {
		if strings.Contains(text, m) {
			return true
		}
	}
	return false
}

func looksLikeCode(text string) bool {
	if strings.Contains(text, "```") {
		return true
	}
	codeLines := 0
	for _, line := range strings.Split(text, "\n") {
		stripped := strings.TrimLeft(line, " \t\n\r\f\v")
		for _, p := range codeLinePrefixes {
			if strings.HasPrefix(stripped, p) {
				codeLines++
				break
			}
		}
	}
	return codeLines >= minCodeLines
}

// looksLikeStructured: JSON-key density over code points — procedural scan
// for `"key":` shapes (1-64 code points between quotes, optional spaces or
// tabs before the colon).
func looksLikeStructured(text string) bool {
	chars := []rune(text)
	n := len(chars)
	if n < structuredMinChars {
		return false
	}
	keys := countJSONKeys(chars)
	// keys / (n/1000) >= 1.0  ⇔  keys * 1000 >= n
	return keys*1000 >= n
}

func countJSONKeys(chars []rune) int {
	count := 0
	i := 0
	for i < len(chars) {
		if chars[i] != '"' {
			i++
			continue
		}
		start := i + 1
		j := start
		ok := false
		for j < len(chars) && j-start <= 64 {
			if chars[j] == '"' {
				ok = j > start
				break
			}
			if chars[j] == '\n' {
				break
			}
			j++
		}
		if !ok {
			i++
			continue
		}
		k := j + 1
		for k < len(chars) && (chars[k] == ' ' || chars[k] == '\t') {
			k++
		}
		if k < len(chars) && chars[k] == ':' {
			count++
			i = k + 1
		} else {
			i = j + 1
		}
	}
	return count
}

// RepeatTracker flags duplicate message content across a session. Not safe
// for concurrent use — one per session.
type RepeatTracker struct {
	seen map[string]struct{}
}

func NewRepeatTracker() *RepeatTracker {
	return &RepeatTracker{seen: make(map[string]struct{})}
}

// IsRepeat reports whether this message's content was seen before, and
// records it as seen.
func (t *RepeatTracker) IsRepeat(message Message) bool {
	h, ok := repeatHash(message)
	if !ok {
		return false
	}
	if _, dup := t.seen[h]; dup {
		return true
	}
	t.seen[h] = struct{}{}
	return false
}

func (t *RepeatTracker) Reset() {
	t.seen = make(map[string]struct{})
}

func repeatHash(message Message) (string, bool) {
	// tool_use/tool_result blocks pair by id; dropping a "duplicate"
	// tool_result would orphan its tool_use. Never hash those messages.
	if hasToolLinkage(message) {
		return "", false
	}
	text := GetTextContent(message)
	if text == "" {
		return "", false
	}
	role := strField(message, "role")
	sum := sha256.Sum256([]byte(role + "|" + text))
	return hex.EncodeToString(sum[:]), true
}

func hasToolLinkage(message Message) bool {
	items, ok := message["content"].([]any)
	if !ok {
		return false
	}
	for _, item := range items {
		if b, ok := item.(map[string]any); ok {
			t := strField(b, "type")
			if t == "tool_use" || t == "tool_result" {
				return true
			}
		}
	}
	return false
}

File: rust/src/classifier.rs
//! Content-type classifier + RepeatTracker (parity-spec §6).

use std::collections::HashSet;

use crate::content::get_text_content;
use crate::json::Value;
use crate::sha256;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ContentType {
    Unknown,
    Prose,
    Code,
    Error,
    Structured,
    Repeat,
    LongImportant,
}

impl ContentType {
    pub fn as_str(&self) -> &'static str {
        match self {
            ContentType::Unknown => "unknown",
            ContentType::Prose => "prose",
            ContentType::Code => "code",
            ContentType::Error => "error",
            ContentType::Structured => "structured",
            ContentType::Repeat => "repeat",
            ContentType::LongImportant => "long_important",
        }
    }

    pub fn parse(s: &str) -> Option<ContentType> {
        Some(match s {
            "unknown" => ContentType::Unknown,
            "prose" => ContentType::Prose,
            "code" => ContentType::Code,
            "error" => ContentType::Error,
            "structured" => ContentType::Structured,
            "repeat" => ContentType::Repeat,
            "long_important" => ContentType::LongImportant,
            _ => return None,
        })
    }
}

const ERROR_MARKERS: &[&str] = &[
    "Traceback (most recent call last):",
    "Uncaught exception",
    "UnhandledPromiseRejection",
    "thread 'main' panicked at",
    "panic: ",
    "Exception in thread",
    "FATAL: ",
    "ERROR: ",
    "Error: ",
    "Exception: ",
    "java.lang.",
];

const CODE_LINE_PREFIXES: &[&str] = &[
    "def ", "class ", "function ", "async function ", "import ", "from ",
    "export ", "package ", "#include", "fn ", "pub fn ", "func ", "var ",
    "const ", "let ",
];

const MIN_CODE_LINES: usize = 2;
const STRUCTURED_MIN_CHARS: usize = 200;

/// Classify a message: ERROR > CODE > STRUCTURED > PROSE; UNKNOWN when the
/// extracted text is empty.
pub fn classify(message: &Value) -> ContentType {
    let text = get_text_content(message);
    if text.trim().is_empty() {
        return ContentType::Unknown;
    }
    if looks_like_error(&text) {
        return ContentType::Error;
    }
    if looks_like_code(&text) {
        return ContentType::Code;
    }
    if looks_like_structured(&text) {
        return ContentType::Structured;
    }
    ContentType::Prose
}

fn looks_like_error(text: &str) -> bool {
    ERROR_MARKERS.iter().any(|m| text.contains(m))
}

fn looks_like_code(text: &str) -> bool {
    if text.contains("```") {
        return true;
    }
    let code_lines = text
        .split('\n')
        .filter(|line| {
            let stripped = line.trim_start();
            CODE_LINE_PREFIXES.iter().any(|p| stripped.starts_with(p))
        })
        .count();
    code_lines >= MIN_CODE_LINES
}

/// JSON-key density: procedural scan for `"key":` shapes (1–64 code points
/// between quotes, optional spaces/tabs before the colon).
fn looks_like_structured(text: &str) -> bool {
    let chars: Vec<char> = text.chars().collect();
    let n = chars.len();
    if n < STRUCTURED_MIN_CHARS {
        return false;
    }
    let keys = count_json_keys(&chars);
    // keys / (n/1000) >= 1.0  ⇔  keys * 1000 >= n
    keys * 1000 >= n
}

fn count_json_keys(chars: &[char]) -> usize {
    let mut count = 0;
    let mut i = 0;
    while i < chars.len() {
        if chars[i] != '"' {
            i += 1;
            continue;
        }
        // Find the closing quote within 1..=64 code points, no " or \n inside.
        let start = i + 1;
        let mut j = start;
        let mut ok = false;
        while j < chars.len() && j - start <= 64 {
            match chars[j] {
                '"' => {
                    ok = j > start; // at least one code point inside
                    break;
                }
                '\n' => break,
                _ => j += 1,
            }
        }
        if !ok {
            i += 1;
            continue;
        }
        // Optional spaces/tabs, then ':'
        let mut k = j + 1;
        while k < chars.len() && (chars[k] == ' ' || chars[k] == '\t') {
            k += 1;
        }
        if k < chars.len() && chars[k] == ':' {
            count += 1;
            i = k + 1;
        } else {
            // Overlapping matches are impossible in the reference regex scan:
            // resume after the closing quote.
            i = j + 1;
        }
    }
    count
}

/// Tracks content hashes across a session to flag duplicate messages.
#[derive(Default)]
pub struct RepeatTracker {
    seen: HashSet<String>,
}

impl RepeatTracker {
    pub fn new() -> Self {
        Self::default()
    }

    /// True if this message's content was seen before; records it as seen.
    pub fn is_repeat(&mut self, message: &Value) -> bool {
        let h = Self::hash(message);
        let Some(h) = h else { return false };
        if self.seen.contains(&h) {
            return true;
        }
        self.seen.insert(h);
        false
    }

    pub fn reset(&mut self) {
        self.seen.clear();
    }

    fn hash(message: &Value) -> Option<String> {
        // tool_use/tool_result blocks pair by id; dropping a "duplicate"
        // tool_result would orphan its tool_use. Never hash those messages.
        if has_tool_linkage(message) {
            return None;
        }
        let text = get_text_content(message);
        if text.is_empty() {
            return None;
        }
        let role = message.str_field("role");
        Some(sha256::hex_digest(format!("{role}|{text}").as_bytes()))
    }
}

pub fn has_tool_linkage(message: &Value) -> bool {
    let Some(items) = message.get("content").and_then(Value::as_arr) else {
        return false;
    };
    items.iter().any(|b| {
        matches!(b.str_field("type"), "tool_use" | "tool_result")
    })
}
